numpy/random/mtrand/mtrand.pyx

```diff
@@ -5064,7 +5064,8 @@ cdef class RandomState:
             x_ptr = <char*><size_t>x.ctypes.data
             stride = x.strides[0]
             itemsize = x.dtype.itemsize
-            buf = np.empty_like(x[0])  # GC'd at function exit
+            # Ensure that buf has the same dtype as x
+            buf = np.empty_like(x[:1])  # GC'd at function exit
```

Comment:
Bit subtle, but `buf` is left containing a reference to some object in `x`, which is decrefed when the buffer is destroyed. Furthermore, `buf` begins containing a reference to `None`, so there is also a memory leak. The only other fix I can see to define the buffer as `int8` with length `itemsize`.